include/envoy/network/listen_socket.h

```diff
@@ -33,30 +35,34 @@ class Socket {
   /**
    * Visitor class for setting socket options.
    */
-  class Options {
+  class Option {
   public:
-    virtual ~Options() {}
+    virtual ~Option() {}
 
     /**
      * @param socket the socket on which to apply options.
+     * @param pre_bind 'true' if socket has has not been bound yet, and is going to be bound
+     *        after the options are set. This an be used to determine whether to set options
+     *        that only make sense for bound sockets if set before the bind() call.
      * @return true if succeeded, false otherwise.
      */
-    virtual bool setOptions(Socket& socket) const PURE;
+    virtual bool setOption(Socket& socket, bool pre_bind) const PURE;
 
     /**
-     * @return bits that can be used to separate connections based on the options. Should return
-     *         zero if connections with different options can be pooled together. This is limited
-     *         to 32 bits to allow these bits to be efficiently combined into a larger hash key
-     *         used in connection pool lookups.
+     * @param vector of bits that can be used to separate connections based on the options. Should
+     *        return zero if connections with different options can be pooled together. This is
+     *        limited to 32 bits to allow these bits to be efficiently combined into a larger hash
+     *        key used in connection pool lookups.
      */
-    virtual uint32_t hashKey() const PURE;
+    virtual void hashKey(std::vector<uint8_t>& key) const PURE;
   };
-  typedef std::shared_ptr<Options> OptionsSharedPtr;
+  typedef std::shared_ptr<Option> OptionSharedPtr;
+  typedef std::shared_ptr<std::list<OptionSharedPtr>> OptionsSharedPtr;
```

Comment:
std::vector is better than std::list for this, I think.

Reply:
Will change.

Reply:
Changed.